### PR TITLE
TRLE Colours Mode

### LINKED ISSUE #49: TRLE colours mode
The program needs a way of temporarily disabling the original textures and instead texture the level using TRLE colours (blue floor, green walls, pink triggers etc). This will make it much easier to discern squares and triggers.

- There should be an option in the view menu to toggle TRLE colours
- When selected, texture rendering should be disabled and replaced with TRLE colours

Notes:
Hidden geometry should be rendered, as it would be seen if it was the TRLE editor.
Walls are not currently rendered, so they'll have to be added when they're added.

- [x] Add checkbox to view options to toggle TRLE colours mode
- [x] Override colours for level mesh somehow
- [ ] Add tests
- [x] Fix walls on portals
- [x] Fix ceilings on walls
- [x] Fix big columns
- [ ] Triangles are in the wrong direction on the floor (eg - SW -> NE instead of SE -> NW)
- [x] Remove duplicate triangles
- [x] Add grid instead of chequerboard
- [x] Ceiling triangulation in TR3+
- [x] Floor triangulation in TR3+ (diagonal half tiles)
- [x] Use TRLE mesh for collision mesh

### LINKED ISSUE #600: Show all geometry
Render all collision geometry, including invisible walls around the edge of rooms. Need to draw the side of them, not just the top.

### PR DESCRIPTION
**Pull request by chreden**
Add TRLE colours mode.
This renders the geometry derived from the floordata instead of the visible geometry.
Closes #49
Closes #600

### CODE REVIEW

File: trview.app/Application.cpp [hunk 1]
```diff
@@ -155,6 +155,8 @@ namespace trview
         _level = _level_source(std::move(new_level));
         _level->set_filename(filename);
 
+        _level->set_map_colours(_settings.map_colours);
```
Comment (chreden) [inline trview.app/Application.cpp:158]: Add test

File: trview.app/Application.cpp [hunk 2]
```diff
@@ -305,6 +307,10 @@ namespace trview
             _route_window->set_randomizer_enabled(settings.randomizer_tools);
             _route->set_randomizer_enabled(settings.randomizer_tools);
             _rooms_windows->set_map_colours(settings.map_colours);
+            if (_level)
+            {
+                _level->set_map_colours(settings.map_colours);
```
Comment (chreden) [inline trview.app/Application.cpp:312]: Test that this happens

File: trview.app/Elements/IRoom.h [hunk 3]
```diff
@@ -253,6 +261,34 @@ namespace trview
...
+        /// Gets the y bottom value of the room.
+        /// </summary>
+        /// <returns></returns>
+        virtual float y_bottom() const = 0;
+        /// <summary>
+        /// Gets the y top value of the room.
+        /// </summary>
+        /// <returns></returns>
+        virtual float y_top() const = 0;
+        /// <summary>
```
Comment (chreden) [inline trview.app/Elements/IRoom.h:274]: Fix comments

File: trview.app/Elements/ILevel.h [hunk 4]
```diff
@@ -104,5 +109,6 @@ namespace trview
         /// Event raised when something has changed in the appearance of the level or the
         /// items that are contained within.
         Event<> on_level_changed;
+        mutable Event<> on_trle_colours_changed;
```
Comment (chreden) [inline trview.app/Elements/ILevel.h:112]: Try and find a way to not do this.

File: trview.app/Elements/ISector.cpp [hunk 5]
```diff
@@ -1,8 +1,103 @@
...
+
+    ISector::Triangle::Triangle(const Vector3& v0, const Vector3& v1, const Vector3& v2, SectorFlag type, uint32_t room)
+        : Triangle(v0, v1, v2, Vector2::Zero, Vector2::Zero, Vector2::Zero, type, room)
+    {
+    }
+
+    ISector::Quad::Quad(const Vector3& v0, const Vector3& v1, const Vector3& v2, const Vector3& v3, SectorFlag type, uint32_t room)
+        : v{ v0, v1, v2, v3 }, uv{ Vector2(0, 1), Vector2(1, 0), Vector2(0, 0), Vector2(1, 1) }, type(type), room(room)
+    {
+        // Calculate a bunch of stats that can be used later.
```
Comment (chreden) [inline trview.app/Elements/ISector.cpp:71]: Useless comment

Comment (chreden) [inline trview.app/Elements/ISector.cpp:71]: Useless comment

**Review by makotocchi — APPROVED** [review]

**Pull request merged**